Add waitForProcessingState to engine

remcowesterhoud opened the pull request:
## Description

<!-- Please explain the changes you made here. -->
In #202 we have a flaky tests. In this test we have a process that's triggered by a timer start event. We increase the time of the engine to trigger this timer, sleep for 100 ms, wait for the engine to reach an idle state and assert that the process instance gets completed.

This solution has a few problems. First of all we use a `Thread.sleep()` to give the engine some time to trigger the timers. This is a dirty solution and it'd be better if we have a solution that only waits until the engine is running again.
The second problem is that the engine has not always triggered the timers within 100 ms. Because of this the engine is considered idle, without actually finishing the processing.

This PR adds a `waitForProcessingState` to the engine. This method is very similar to the `waitForIdleState` method we already have. Just like that method it uses the `IdleStateMonitor` with a callback to wait until the engine has started committing records again.
A timeout needs to be specified to prevent waiting forever in case the engine will be permanently idle.

## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #202 

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda-cloud/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to backport the fix

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually

Documentation:
* [ ] Javadoc has been written
* [ ] The documentation is updated

pihme commented on api/src/main/java/io/camunda/zeebe/process/test/api/InMemoryEngine.java:
```diff
@@ -29,10 +30,15 @@
+   * @throws TimeoutException if the engine has not reached a processing state before the timeout
    */
-  void waitForIdleState();
+  void waitForProcessingState(Duration timeout) throws InterruptedException, TimeoutException;
```
:wrench: 
```suggestion
  void waitForBusyState(Duration timeout) throws InterruptedException, TimeoutException;
```

Maybe also change order of methods:
1. Wait for Busy
2. Wait for Idle

pihme replied:
:wrench: Maybe harmonize interfaces and also add a timeout to `waitForIdleState`

pihme replied:
btw, totally fine to keep the name processing state in the idle state monitor, because there it is more technical and closer to the implementation. But here I have a slight preference for using the busy idle pair. But it's only a mild preference, so I am also ok with sticking with the name.

pihme commented on engine/src/main/java/io/camunda/zeebe/process/test/engine/IdleStateMonitor.java:
```diff
@@ -44,9 +45,16 @@
+    synchronized (idleCallbacks) {
+      idleCallbacks.add(callback);
+    }
+    checkIdleState();
```
:x: rename method to something like `checkEngineState` or `checkProcessingOrIdleState` or something along those lines.

Maybe we should also rename the method to `checkStateAndNotifyCallbacks` to hint to the side effect

pihme replied:
:x: dito for renaming the class altogether

pihme commented on qa/src/test/java/io/camunda/zeebe/process/test/qa/util/Utilities.java:
```diff
@@ -149,15 +150,16 @@ public static PublishMessageResponse sendMessage(
-      Thread.sleep(100);
-    } catch (InterruptedException e) {
-      // do nothing
+      engine.waitForProcessingState(Duration.ofSeconds(1));
```
:wrench: maybe also add a `waitForProcessingState` method to this class so the two lines look more alike

pihme reviewed (changes requested):
Looks good!

Some suggestions for name changes.

And maybe we should start writing tests also for classes like the idle state. The longer we wait, the more painful it will be to start writing tests for a class like that.

remcowesterhoud commented:
Thanks for reviewing @pihme. I've applied your suggestions and created a unit test for the `IdleStateMonitor`. The test is somewhat complex because I needed to make my own `LogStreamReader`. If you see a better solution for this please let me know!

pihme commented on engine/src/main/java/io/camunda/zeebe/process/test/engine/EngineStateMonitor.java:
```diff
@@ -49,22 +50,22 @@ public void addOnIdleCallback(final Runnable callback) {
       } else {
-        cancelNotification();
+        cancelIdleStateNotification();
         runProcessingCallbacks();
```
🔧 maybe rename to `notifyProcessingCallbacks`

pihme commented on engine/src/test/java/io/camunda/zeebe/process/test/engine/EngineStateMonitorTest.java:
```diff
@@ -0,0 +1,178 @@
+class EngineStateMonitorTest {
+
+  @Test
+  void testCallbackIsCalledWhenEngineIsAlreadyIdle()
```
:x: please qualify which callback is being called in test methods names (dito for the following)

pihme commented on engine/src/test/java/io/camunda/zeebe/process/test/engine/EngineStateMonitorTest.java:
```diff
@@ -0,0 +1,178 @@
+    final InMemoryLogStorage logStorage = mock(InMemoryLogStorage.class);
+    final TestLogStreamReader logStreamReader = new TestLogStreamReader();
+    final EngineStateMonitor monitor = new EngineStateMonitor(logStorage, logStreamReader);
+    changeToBusyState(monitor, logStreamReader);
```
🤔 I wonder what would be clearer to read. We do the setup of the callbacks in _given_ and then in _when_ we change to busy state. Or we do it like it is now. But, personally, I find this hard to read. Because the when/then part is _when I register a callback, it is being called_. 

I like this test implementation, because it has an explicit call to _changeToBusyState_ I wonder whether we could also manufacture a _changeToIdleState_ method alongside and use it in the test case above.

pihme replied:
Ah, found the _changeToIdleState_ method.

remcowesterhoud replied:
It's hard to structure the given/when/then like you suggested. This is because when we add a callback to the `EngineStateMonitor` it immediately checks if the engine is busy or idle, and then notifies the callbacks if necessary. To circumvent this we need to call `changeToBusyState` or `changeToIdleState` before we actually add the callback.

I will make some other changes to make it more readable and less confusing though.

pihme commented on engine/src/test/java/io/camunda/zeebe/process/test/engine/EngineStateMonitorTest.java:
```diff
@@ -0,0 +1,178 @@
+    reader.setLastEventPosition(position);
+  }
+
+  private class TestLogStreamReader implements LogStreamReader {
```
:wrench: I think you could just use mocks for this. But I don't think this too verbose, either.

remcowesterhoud replied:
I'd rather stick with this solution in that case. I think mocking is gonna make the test more confusing.

pihme reviewed (approved):
Looks good!

I have some comments. 

If you make bigger changes to the test, I would like to review those again. Otherwise, no need to do another review round

remcowesterhoud commented:
@pihme I made some bigger changes to the test so I'd like another review.

The biggest change is that I added a way to lock the state. This is required because when we switch to a busy state and register a callback the check if we are already in an idle state. This executes this method:

```java
private void forwardToLastEvent() {
    synchronized (reader) {
      while (reader.hasNext()) {
        lastEventPosition = reader.next().getPosition();
      }
    }
  }
```

This iteration causes the state to become idle. Before this change as a workaround I registered the busy callback before actually changing the state to busy. This was confusing so I switched this up, but because of it I required a way to lock the state into a busy state. That's why I added the `stateLocked` flag.

pihme commented on engine/src/test/java/io/camunda/zeebe/process/test/engine/EngineStateMonitorTest.java:
```diff
@@ -161,7 +165,7 @@ public void close() {}
     @Override
     public boolean hasNext() {
-      return position < lastEventPosition;
+      return !stateLocked && position < lastEventPosition;
```
:x: I think this is a bit misleading in terms of naming and what it does. It seems to lock only one of the states. 

I think it would be clearer to either rename `stateLocked` or to capture the state at the moment it is locked:
````
setStateLocked(stateLocked) {
  this.stateLocked = stateLocked;
  if (stateLocked) {
     capturedState = position < lastPosition; 
  }
}

hasNext()  {
 return stateLocked ? capturedState : position < lastPosition
}
````

remcowesterhoud replied:
I'm not sure what you mean here. `stateLocked` definitely locks both of the states.

If the state is locked `!stateLocked && position < lastEventPosition;` will always resolve as `false`. This has as a result that the `EngineStateMonitor` will never update it's position. This means that if the positions are equal, the state is locked to idle. If the positions are not equal the state is locked to busy.
```java
synchronized (reader) {
      while (reader.hasNext()) { // <--- this will resolve to false so the lastEventPosition is never updated.
        lastEventPosition = reader.next().getPosition();
      }
    }
```

pihme replied:
Hmm, I see what you mean. The thing that is confusing me is that `hasNext` is always false when `stateLocked=true`. Which seems wrong, in my mind. And the only thing it does work is indirectly because of taking into account implementation details of how this method is used as a guard around the call to next.

Maybe another way could be to implement it like this:
```
    @Override
    public LoggedEvent next() {
     if !stateLocked {
        position++;
    }
      final LoggedEvent event = mock(LoggedEvent.class);
      when(event.getPosition()).thenReturn(position);
      return event;
    }
```

So then `hasNext` would give always the correct result, but one couldn't really advance. But maybe this then results in endless loops 🤔

remcowesterhoud replied:
Yes, this solution will result in infinite loops.

I don't like the other solution either though:
```java
setStateLocked(stateLocked) {
  this.stateLocked = stateLocked;
  if (stateLocked) {
     capturedState = position < lastPosition; 
  }
}

hasNext()  {
 return stateLocked ? capturedState : position < lastPosition
}
```
This way it seems like. the state is being managed by the reader which is not the case.

The way it is currently makes sense in my mind. The `EngineStateMonitor` asks if there is any more events it should process. Because the state is locked we override this by saying, no there isn't anything. Thus, keeping the state of the `EngineStateMonitor` the same.

pihme replied:
What about we time box this, each of us thinks 30mins hard about the problem. If we find no better solution, we leave it as it is?

remcowesterhoud replied:
Sounds like a plan

pihme replied:
I think this is goo enough for now

pihme reviewed (changes requested):
Quick feedback; didn't do a full review yet

pihme commented on engine/src/test/java/io/camunda/zeebe/process/test/engine/EngineStateMonitorTest.java:
```diff
@@ -75,57 +76,60 @@ void testCallbackIsCalledWhenEngineIsAlreadyBusy()
 
   @Test
-  void testCallbackIsCalledWhenBusyStateIsReached()
+  void testOnProcessingCallbackIsCalledWhenBusyStateIsReached()
```
:x: If I am not mistaken this test now has exactly the same code as `...IsAlreadyBusy`

pihme commented on engine/src/test/java/io/camunda/zeebe/process/test/engine/EngineStateMonitorTest.java:
```diff
@@ -10,40 +10,48 @@
     final Runnable callback = () -> callbackFuture.complete(null);
+
+    // when
+    changeToBusyState(monitor, logStreamReader, false);
```
It's not so clear to see which line in the when part is responsible for reaching the idle state.

pihme reviewed (changes requested):
Definitely going in the right direction, Some details are still confusing

pihme reviewed (approved):
🎉
